### PR TITLE
First open_esplibs batch, libmain (most of libmain open sourced)

### PR DESCRIPTION
**Pull request by projectgus**
@foogod has done great work with this and I think now is a good time to merge it. I originally wanted to get automated tests running firsts, but that's been too long coming (my fault) and there are issues (#147, possibly #80) that it covers.

I've tested many of the more complex examples (http_get_mbedtls, ota_basic, mqtt-client) with this and they all seem fine. There may be more subtle bugs introduced, but I think the only way to find them will be if people start using these changes.

@foogod has made an excellent set of compile-time selection switches, so if there are bugs they should be easy to isolate.

### CODE REVIEW

**Comment by projectgus** [comment]
Hmm, I've done something wrong with the merge here - it's pulled in a bunch of double commits ie 36886412 vs b0844b01.

@foogod, if I rebase this whole branch against current master then will that make things hard for you in any way?

File: core/include/esp/sar_regs.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+/* Note: This memory region is not currently well understood.  Pretty much all
+ * of the definitions here are from reverse-engineering the Espressif SDK code,
+ * many are just educated guesses, and almost certainly some are misleading or
+ * wrong.  If you can improve on any of this, please contribute!
+ */
+
+struct SAR_REGS {
+    uint32_t volatile _unknown0[18];   // 0x00 - 0x44
+    uint32_t volatile UNKNOWN_48;      // 0x48 : used by sdk_system_restart_in_nmi()
```
Comment (pfalcon) [inline core/include/esp/sar_regs.h]: These two lines show how inconsistent naming is, leading to stuff like "unknownc" above. Can there be convention like "unkown_0c" (underscore, 2 digits)?

Comment (foogod) [inline core/include/esp/sar_regs.h]: These are not inconsistant.  "_unknown" is explicitly intended to be a temporary placeholder for spacing purposes and should not be referenced by any code (there may be one or two places that do reference them, but if there is that's an error and needs to be cleaned up).

The `UNKNOWN_48` is an example of the (rare) situation where a field needed to be referenced somewhere but I still could not figure out what it should be named from the context.  The uppercase indicates it's not just a spacer/placeholder but is actually a referenced field.

If you want to go through and clean up the "_unknown" names with underscores/leading digits/etc, you certainly can, but frankly it seems pretty minor to me, considering that they are explicitly not supposed to be being referenced by anything (but if somebody wants to do a PR for that that's fine by me)..

**Comment by foogod** [comment]
Feel free to rebase the branch (I don't think it'll screw up anything for me at this point, but if it does, I can deal with it :) )

**Comment by projectgus** [comment]
Thanks, rebased. This looks ready to merge to me.

File: open_esplibs/libmain/timers.c [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    new_entry->next = 0;
+    tailptr = &entry->next;
+    if (!timer_list) {
+        tailptr = &timer_list;
+    }
+    *tailptr = new_entry;
+}
+
+void sdk_os_timer_arm(ETSTimer *ptimer, uint32_t milliseconds, bool repeat_flag) {
+    if (!ptimer->timer_handle) {
```
Comment (UncleRus) [inline open_esplibs/libmain/timers.c]: Bug in condition

**Pull request merged**